cunumeric/deferred.py

```diff
@@ -290,7 +333,7 @@ def conj(self):
 
     # Copy source array to the destination array
     @auto_convert([1])
-    def copy(self, rhs, deep=False):
+    def copy(self, rhs: Any, deep: bool = False) -> None:
```

Comment:
The body of `DeferredArray` methods is written assuming that all `NumPyThunk`-type arguments are actually `DeferredArray`s. The `@auto_convert` decorator converts incoming `EagerArray` arguments to `DeferredArray`, thus allowing the method to handle any `NumPyThunk`. It would, therefore, be most accurate to write the function signature as:

```
copy(self, rhs: DeferredArray, deep: bool = False) -> None
```

meaning that mypy can analyze the method knowing that `rhs` will always be of type `DeferredArray`, as the code assumes.

However, we also need to somehow inform mypy that the presence of `@auto_convert([1])` changes the type signature visible to the outside world to:

```
copy(self, rhs: NumPyThunk, deep: bool = False) -> None
```

If we just change the type signature to use `DeferredArray`, then mypy will not reason about what the decorator is doing, and will end up complaining that the method's signature doesn't match that on the base class (which claims to accept any `NumPyThunk` object).

Reply:
Currenty `auto_convert` specifies this
```
R = TypeVar("R")
P = ParamSpec("P")

def auto_convert(
    indices: Collection[int], keys: Sequence[str] = []
) -> Callable[[Callable[P, R]], Callable[P, R]]:
```

`ParamSpec` is actually quite new, I will have to see there is any way to take apart and re-assemble just certain pieces of it in the return.  I'm pretty sure this kind of type surgery would be possible in TypeScript, but I am a bit skeptical that `auto_convert` can be improved much, unless it can also be simplified some.

Reply:
Other approaches that would certainly work but are a bit more verbose:

* change `auto_convert` to be a helper function rather than a decorator and use it inside the method explicitly
* optionally split methods into `_copy` (takes Deferred) and `copy` (that takes NumpyThunk) that only calls the helper and then delegates to `_copy`.

This would be a little tedious but I think ultimately a little clearer,and there are only ~20 uses of `auto_convert`.